tracer/src/Datadog.Trace/Ci/TestModule.cs

```diff
@@ -0,0 +1,315 @@
+    public void SetTag(string key, double? value)
+    {
+        _span.SetMetric(key, value);
+    }
+
+    /// <summary>
+    /// Set Error Info
+    /// </summary>
+    /// <param name="type">Error type</param>
+    /// <param name="message">Error message</param>
+    /// <param name="callStack">Error callstack</param>
+    public void SetErrorInfo(string type, string message, string? callStack)
+    {
+        var span = _span;
+        span.Error = true;
+        span.SetTag(Trace.Tags.ErrorType, type);
+        span.SetTag(Trace.Tags.ErrorMsg, message);
+        if (callStack is not null)
+        {
+            span.SetTag(Trace.Tags.ErrorStack, callStack);
+        }
+    }
+
+    /// <summary>
+    /// Set Error Info from Exception
+    /// </summary>
+    /// <param name="exception">Exception instance</param>
+    public void SetErrorInfo(Exception exception)
+    {
+        _span.SetException(exception);
+    }
+
+    /// <summary>
+    /// Close test module
+    /// </summary>
+    public void Close()
+    {
+        Close(null);
+    }
+
+    /// <summary>
+    /// Close test module
+    /// </summary>
+    /// <param name="duration">Duration of the test module</param>
+    public void Close(TimeSpan? duration)
+    {
+        if (Interlocked.Exchange(ref _finished, 1) == 1)
+        {
+            return;
+        }
+
+        var span = _span;
+
+        // Calculate duration beforehand
+        duration ??= span.Context.TraceContext.ElapsedSince(span.StartTime);
+
+        var remainingSuites = Array.Empty<TestSuite>();
+        lock (_suites)
+        {
+            if (_suites.Count > 0)
+            {
+                remainingSuites = _suites.Values.ToArray();
+            }
+        }
+
+        foreach (var suite in remainingSuites)
+        {
+            suite.Close();
+        }
+
+        // Update status
+        Tags.Status ??= TestTags.StatusPass;
+
+        // Finish if agentless is enabled
+        if (CIVisibility.Settings.Agentless)
+        {
+            span.Finish(duration.Value);
+        }
+
+        Current = null;
+        CIVisibility.Log.Debug("### Test Module Closed: {name}", Name);
+        CIVisibility.FlushSpans();
+    }
+
+    /// <summary>
+    /// Create a new test suite for this session
+    /// </summary>
+    /// <param name="name">Name of the test suite</param>
+    /// <returns>Test suite instance</returns>
+    public TestSuite CreateSuite(string name)
+    {
+        return CreateSuite(name, null);
+    }
+
+    /// <summary>
+    /// Create a new test suite for this session
+    /// </summary>
+    /// <param name="name">Name of the test suite</param>
+    /// <param name="startDate">Test suite start date</param>
+    /// <returns>Test suite instance</returns>
+    public TestSuite CreateSuite(string name, DateTimeOffset? startDate)
+    {
+        var suite = new TestSuite(this, name, startDate);
+        lock (_suites)
+        {
+            _suites[name] = suite;
```

Comment:
This could overwrite an existing suite if we're not careful (or since this is a public API, if the user is not careful). Should we first try calling `GetSuite` and returning that result if there is one?

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/commit/1804c381ada9ab0c3c2fcbab2e47173544f57df9